File: karatsuba_multiplication.js
/*
x*y = 10^n * (ac) + 10^n/2 * (ad + bc) + bd
In a previous example with 4-digit numbers, there were 4 recursive calls to compute the 4 products necessary for computing the final product.
However, since we simply care about the sum of ad and bc, we can do this in 3 operations for Karatsuba’s algorithm:
1. Recursively compute ac
2. Recursively compute bd
3. Recursively compute the product of (a + b)(c + d) = ac + ad + bc + bd

Assumes x and y are of equal length
Assumes x and y have a number of digits that is a power of 2
*/
function karatsuba(x, y, len) {
  var xStr = x.toString();
  var yStr = y.toString();
  if (len <= 2) return x*y;

  var halfLen = Math.floor(len / 2);
  var a = Number(xStr.slice(0, halfLen));
  var b = Number(xStr.slice(halfLen));
  var c = Number(yStr.slice(0, halfLen));
  var d = Number(yStr.slice(halfLen));

  var ac = karatsuba(a, c, halfLen);
  var bd = karatsuba(b, d, halfLen);
  var sumAdBc = karatsuba(a+b, c+d, halfLen) - bd - ac;

  return Math.pow(10,len)*ac + Math.pow(10,halfLen)*sumAdBc + bd;
}

var x = 8576456798574326;
var y = 7465358906879078;
console.log('correct: ', x*y);
console.log('karatsuba: ', karatsuba(x, y, 16));

File: karatsuba_multiplication.py
# Assumes x and y are of equal length
# Assumes x and y have a number of digits that is a power of 2

def karatsuba(x, y, length):
  a = x // 10**(length//2)9
  b = x % 10**(length//2)
  c = y // 10**(length//2)
  d = y % 10**(length//2)

  if (length == 2):
    return x*y

  ac = karatsuba(a, c, length//2)
  bd = karatsuba(b, d, length//2)
  sumAdBc = karatsuba(a+b, c+d, length//2) - bd - ac
  
  return 10**length*ac + 10**(length//2)*sumAdBc + bd
  
  
  
x = 3141592653589793238462643383279502884197169399375105820974944592;
y = 2718281828459045235360287471352662497757247093699959574966967627;
print('1: ', x*y);
print('2: ', karatsuba(x, y, 16));
